File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/linear_discriminant_analysis_tutorial.m
close all; clear all; clc; 

% Generate synthetic data for two classes
rng(1); % For reproducibility
N = 100;
X1 = randn(N,2) + 2;
X2 = randn(N,2) - 2;
X = [X1; X2];
y = [ones(N,1); 2*ones(N,1)];

[W, Z] = lda(X, y);

% Plot original data and projected data
figure;

% --------- Subplot 1: Original Data ---------
subplot(1,2,1);
scatter(X1(:,1), X1(:,2), 'r', 'filled'); hold on;
scatter(X2(:,1), X2(:,2), 'b', 'filled');
title('Original Data');
xlabel('Feature 1'); ylabel('Feature 2'); grid on;

% Draw a dashed line representing the first LDA direction
% 1) Compute the mean of all data
mu = mean(X,1);  % 1x2

% 2) Take the first column of W as the primary discriminant direction
dirVec = W(:,1);  % 2x1

% 3) Generate a set of t-values to plot the line across a suitable range
t = linspace(-80, 80, 100);

% 4) Parametric equation of the line: point(t) = mu + t*dirVec
linePoints = mu + t' * dirVec';  % 100x2

% 5) Plot the dashed line
plot(linePoints(:,1), linePoints(:,2), 'k--', 'LineWidth', 2);

% --------- Subplot 2: Projected Data ---------
subplot(1,2,2);
% For visualization, project data onto the first discriminant axis
scatter(Z(1:N,1), zeros(N,1), 'r', 'filled'); hold on;
scatter(Z(N+1:end,1), zeros(N,1), 'b', 'filled');
title('Data Projected onto First Discriminant');
xlabel('Projection Value'); grid on;

% save_all_figs_OPTION('results/linear_discriminant_analysis','png',1)

% --------- LDA Function ---------
function [W, projectedData] = lda(X, y)
    % lda - Linear Discriminant Analysis implementation
    % Inputs:
    %   X - n x d data matrix
    %   y - n x 1 vector of class labels
    % Outputs:
    %   W - Projection matrix
    %   projectedData - Data projected onto the new subspace

    [n, d] = size(X);
    classes = unique(y);
    C = length(classes);

    % Compute overall mean
    mu = mean(X, 1);

    % Initialize scatter matrices
    Sw = zeros(d, d);
    Sb = zeros(d, d);

    for i = 1:C
        Xi = X(y == classes(i), :);
        Ni = size(Xi, 1);
        mu_i = mean(Xi, 1);
        
        % Within-class scatter
        Sw = Sw + (Xi - mu_i)' * (Xi - mu_i);
        
        % Between-class scatter
        diff = (mu_i - mu);
        Sb = Sb + Ni * (diff' * diff);
    end

    % Solve the generalized eigenvalue problem
    [V, D] = eig(Sb, Sw);
    
    % Sort eigenvectors by descending eigenvalues
    [~, idx] = sort(diag(D), 'descend');
    W = V(:, idx);
    
    % Project the data
    projectedData = X * W;
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/linear_discriminant_analysis_tutorial.py
import numpy as np
import matplotlib.pyplot as plt

def lda(X, y):
    """
    Linear Discriminant Analysis (LDA).
    
    Parameters:
      X : n x d data matrix.
      y : n x 1 vector of class labels.
      
    Returns:
      W : Projection matrix (d x d), with eigenvectors sorted by descending eigenvalues.
      projectedData : Data projected onto the new subspace.
    """
    n, d = X.shape
    classes = np.unique(y)
    C = len(classes)
    
    # Compute overall mean
    mu = np.mean(X, axis=0)
    
    # Initialize scatter matrices
    Sw = np.zeros((d, d))
    Sb = np.zeros((d, d))
    
    for cl in classes:
        X_i = X[y == cl]
        Ni = X_i.shape[0]
        mu_i = np.mean(X_i, axis=0)
        # Within-class scatter
        Sw += (X_i - mu_i).T @ (X_i - mu_i)
        # Between-class scatter
        diff = (mu_i - mu).reshape(-1, 1)
        Sb += Ni * (diff @ diff.T)
    
    # Solve the generalized eigenvalue problem: Sb * v = lambda * Sw * v
    eigvals, eigvecs = np.linalg.eig(np.linalg.pinv(Sw) @ Sb)
    
    # Sort eigenvectors by descending eigenvalues
    idx = np.argsort(eigvals)[::-1]
    W = eigvecs[:, idx]
    
    # Project the data
    projectedData = X @ W
    return W, projectedData

# Generate synthetic data for two classes
np.random.seed(1)
N = 100
X1 = np.random.randn(N, 2) + 2
X2 = np.random.randn(N, 2) - 2
X = np.vstack([X1, X2])
y = np.concatenate([np.ones(N), 2 * np.ones(N)])

# Perform LDA
W, Z = lda(X, y)

# Plot original data with the first LDA direction
plt.figure(figsize=(12, 5))

# Subplot 1: Original data
plt.subplot(1, 2, 1)
plt.scatter(X1[:, 0], X1[:, 1], color='red', label='Class 1', alpha=0.7)
plt.scatter(X2[:, 0], X2[:, 1], color='blue', label='Class 2', alpha=0.7)
plt.xlabel('Feature 1')
plt.ylabel('Feature 2')
plt.title('Original Data')
plt.grid(True)
# Draw a dashed line representing the first LDA direction
mu = np.mean(X, axis=0)
dirVec = W[:, 0]
t = np.linspace(-80, 80, 100)
linePoints = mu + np.outer(t, dirVec)
plt.plot(linePoints[:, 0], linePoints[:, 1], 'k--', linewidth=2)

# Subplot 2: Projected data (onto the first discriminant)
plt.subplot(1, 2, 2)
# Project data onto first discriminant axis
proj_class1 = Z[:N, 0]
proj_class2 = Z[N:, 0]
plt.scatter(proj_class1, np.zeros_like(proj_class1), color='red', label='Class 1', alpha=0.7)
plt.scatter(proj_class2, np.zeros_like(proj_class2), color='blue', label='Class 2', alpha=0.7)
plt.xlabel('Projection Value')
plt.title('Data Projected onto First Discriminant')
plt.grid(True)
plt.legend()

plt.tight_layout()
plt.show()
